File: medium/19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next: Optional[ListNode] = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        return self.solution1(head, n)

    def solution1(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        start = end = dummy

        for _ in range(n + 1):
            if start:
                start = start.next

        while start:
            start, end = start.next, end.next

        if end:
            end.next = end.next.next if end.next else None

        return dummy.next

File: medium/19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.ts
/**
 * Definition for singly-linked list.
 */

class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    const dummy = new ListNode(0, head)

    let first: ListNode | null = dummy, second: ListNode | null = dummy

    for (let i = 0; i < n + 1; i++) {
        first = first.next!
    }

    while (first) {
        first = first.next || null, second = second.next!
    }

    if (second.next)
        second.next = second.next?.next
    else
        second = null

    return dummy.next
};

export { }
